database/abusescanner.go

```diff
@@ -0,0 +1,429 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	lock "github.com/square/mongo-lock"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const (
+	// AbuseStatusBlocked denotes the blocked status.
+	AbuseStatusBlocked = "BLOCKED"
+
+	// AbuseStatusNotBlocked denotes the not blocked status.
+	AbuseStatusNotBlocked = "NOT BLOCKED"
+
+	// AbuseDefaultTag is the tag used when there are no tags found in the email
+	AbuseDefaultTag = "abusive"
+
+	// collEmails is the name of the collection that contains all email objects
+	collEmails = "emails"
+
+	// collLocks is the name of the collection that contains locks
+	collLocks = "locks"
+
+	// dbName defines the name of the mongo database
+	dbName = "abuse-scanner"
+
+	// lockOwnerName is passed as the 'Owner' when creating a new lock in
+	// the db for tus uploads.
+	lockOwnerName = "Abuse Scanner"
+
+	// lockTTL is the time-to-live in seconds for a lock
+	lockTTL = 300 // 5 minutes
+)
+
+var (
+	// mongoDefaultTimeout is the default timeout for mongo operations that
+	// require a context but where the input arguments don't contain a
+	// context.
+	mongoDefaultTimeout = time.Minute
+
+	// mongoErrNoDocuments is returned when a database operation completes
+	// successfully but it doesn't find or affect any documents.
+	mongoErrNoDocuments = errors.New("no documents in result")
+)
+
+type (
+	// AbuseScannerDB wraps a generic mongo database
+	AbuseScannerDB struct {
+		MongoDB
+		lock.Client
+		staticPortalHostName string
+	}
+
+	// AbuseEmail represent an object in the emails collection.
+	AbuseEmail struct {
+		ID        primitive.ObjectID `bson:"_id"`
+		UID       string             `bson:"email_uid"`
+		UIDRaw    uint32             `bson:"email_uid_raw"`
+		Body      []byte             `bson:"email_body"`
+		From      string             `bson:"email_from"`
+		Subject   string             `bson:"email_subject"`
+		MessageID string             `bson:"email_message_id"`
+
+		// fields set by parser
+		ParsedAt    time.Time   `bson:"parsed_at"`
+		ParseResult AbuseReport `bson:"parseResult"`
+		Parsed      bool        `bson:"parsed"`
+
+		// fields set by blocker
+		BlockedAt   time.Time `bson:"blocked_at"`
+		BlockResult []string  `bson:"blockResult"`
+		Blocked     bool
+
+		InsertedAt time.Time `bson:"inserted_at"`
+		Finalized  bool
+	}
+
+	// AbuseReport contains all information about an abuse report.
+	AbuseReport struct {
+		Skylinks []string
+		Reporter AbuseReporter
+		Sponsor  string
+		Tags     []string
+	}
+
+	// AbuseReporter encapsulates some information about the reporter.
+	AbuseReporter struct {
+		Name         string
+		Email        string
+		OtherContact string
+	}
```

Comment:
These also need bson tags. They get stored in the DB, so they need tags (both AbuseReport and AbuseReporter).